examples/hdf5.py

```diff
@@ -4,99 +4,199 @@
+        self.sampler = sampler
+        self.loadData()
+
     def viewRangeChanged(self):
-        self.updateHDF5Plot()
-        
-    def updateHDF5Plot(self):
-        if self.hdf5 is None:
+        self.loadData(lazy=True)
+
+    def loadData(self, lazy=False):
+        if self.sampler is None:
             self.setData([])
             return
-        
+
         vb = self.getViewBox()
         if vb is None:
             return  # no ViewBox yet
-        
-        # Determine what data range must be read from HDF5
-        xrange = vb.viewRange()[0]
-        start = max(0,int(xrange[0])-1)
-        stop = min(len(self.hdf5), int(xrange[1]+2))
-        
-        # Decide by how much we should downsample 
-        ds = int((stop-start) / self.limit) + 1
-        
-        if ds == 1:
-            # Small enough to display with no intervention.
-            visible = self.hdf5[start:stop]
-            scale = 1
-        else:
-            # Here convert data into a down-sampled array suitable for visualizing.
-            # Must do this piecewise to limit memory usage.        
-            samples = 1 + ((stop-start) // ds)
-            visible = np.zeros(samples*2, dtype=self.hdf5.dtype)
-            sourcePtr = start
-            targetPtr = 0
-            
-            # read data in chunks of ~1M samples
-            chunkSize = (1000000//ds) * ds
-            while sourcePtr < stop-1: 
-                chunk = self.hdf5[sourcePtr:min(stop,sourcePtr+chunkSize)]
-                sourcePtr += len(chunk)
-                
-                # reshape chunk to be integral multiple of ds
-                chunk = chunk[:(len(chunk)//ds) * ds].reshape(len(chunk)//ds, ds)
-                
-                # compute max and min
-                chunkMax = chunk.max(axis=1)
-                chunkMin = chunk.min(axis=1)
-                
-                # interleave min and max into plot data to preserve envelope shape
-                visible[targetPtr:targetPtr+chunk.shape[0]*2:2] = chunkMin
-                visible[1+targetPtr:1+targetPtr+chunk.shape[0]*2:2] = chunkMax
-                targetPtr += chunk.shape[0]*2
-            
-            visible = visible[:targetPtr]
-            scale = ds * 0.5
-            
-        self.setData(visible) # update the plot
-        self.setPos(start, 0) # shift to match starting index
-        self.resetTransform()
-        self.scale(scale, 1)  # scale to match downsampling
 
-        
+        x1, x2 = vb.viewRange()[0]
+        if not lazy or not self.sampler.isRangeValid(x1, x2):
+            x, y = self.sampler[x1:x2]
+            self.setData(x, y)  # update the plot
+
+
+class DataSampler:
+    def __init__(self, x, y, sampleLimit=2500, cacheLimit=10000000, padding=0.3, eager=True):
+        if x is not None and len(x) != len(y):
+            raise ValueError
+
+        self.x = x
+        self.y = y
+        self.sampleLimit = sampleLimit  # at most 4 times this number of samples will be plotted
```

Comment:
Do these settings need to be dynamically changed at runtime?

Reply:
No, and if they are things would break. They should at least be private until that is fixed.

Reply:
I think that this confirms my thoughts that this shouldn't be a class / object then.

Reply:
There is state that needs to live somewhere even if this particular variable doesn't. Please see the recent commit for the most concise representation of this state that I could come up with.